File: java-logger/core/src/main/java/com/logging/sink/FileSink.java
package com.logging.sink;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.BufferedWriter;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.concurrent.locks.ReentrantLock;
import java.util.function.BiConsumer;

/**
 * Append-only NDJSON sink. One JSON envelope per line; each line is
 * self-describing thanks to the embedded `_log_type` so a sidecar tailer can
 * route mixed records.
 *
 * Designed as the "fast local" path for the sidecar pattern: the application
 * pays only the cost of a buffered file write, and a separate process is
 * responsible for shipping bytes to Kafka, NATS JetStream, or object storage.
 *
 * Rotation by size is supported; when the active file exceeds {@code rotateBytes}
 * it is renamed with a {@code .N} suffix and a fresh file is started.
 */
public final class FileSink implements LogSink {

    private static final Logger LOG = LoggerFactory.getLogger(FileSink.class);

    private final Path activeFile;
    private final long rotateBytes;
    private final boolean fsyncOnFlush;
    private final ReentrantLock lock = new ReentrantLock();

    private BufferedWriter writer;
    private long bytesWritten;
    private int rotationIndex;
    private volatile boolean closed;

    public FileSink(Path activeFile) {
        this(activeFile, 64L * 1024L * 1024L, false);
    }

    public FileSink(Path activeFile, long rotateBytes, boolean fsyncOnFlush) {
        this.activeFile = activeFile;
        this.rotateBytes = rotateBytes;
        this.fsyncOnFlush = fsyncOnFlush;
        try {
            if (activeFile.getParent() != null) {
                Files.createDirectories(activeFile.getParent());
            }
            this.writer = openWriter();
            this.bytesWritten = Files.exists(activeFile) ? Files.size(activeFile) : 0L;
        } catch (IOException e) {
            throw new IllegalStateException("Failed to open file sink at " + activeFile, e);
        }
    }

    private BufferedWriter openWriter() throws IOException {
        return Files.newBufferedWriter(
                activeFile,
                StandardCharsets.UTF_8,
                StandardOpenOption.CREATE,
                StandardOpenOption.APPEND);
    }

    @Override
    public String name() {
        return "file(" + activeFile + ")";
    }

    @Override
    public void publish(LogEnvelope envelope, BiConsumer<Boolean, Throwable> callback) {
        if (closed) {
            if (callback != null) callback.accept(false, new IllegalStateException("FileSink is closed"));
            return;
        }
        String line;
        try {
            line = EnvelopeSerializer.toJson(envelope);
        } catch (Throwable t) {
            if (callback != null) callback.accept(false, t);
            return;
        }
        lock.lock();
        try {
            writer.write(line);
            writer.newLine();
            bytesWritten += line.length() + 1;
            if (rotateBytes > 0 && bytesWritten >= rotateBytes) {
                rotate();
            }
            if (callback != null) callback.accept(true, null);
        } catch (IOException e) {
            LOG.error("FileSink write failed", e);
            if (callback != null) callback.accept(false, e);
        } finally {
            lock.unlock();
        }
    }

    private void rotate() throws IOException {
        writer.flush();
        writer.close();
        rotationIndex++;
        Path rotated = activeFile.resolveSibling(activeFile.getFileName() + "." + rotationIndex);
        Files.move(activeFile, rotated);
        writer = openWriter();
        bytesWritten = 0L;
        LOG.info("FileSink rotated active file -> {}", rotated);
    }

    @Override
    public void flush() {
        if (closed) return;
        lock.lock();
        try {
            writer.flush();
            if (fsyncOnFlush) {
                // Best-effort fsync via re-opening with sync option is too expensive;
                // a Channel-level force is preferable but jumping through Channels here
                // would complicate the buffer ownership. Callers who need durable
                // fsync should use a flat-file-with-fsync transport.
            }
        } catch (IOException e) {
            LOG.error("FileSink flush failed", e);
        } finally {
            lock.unlock();
        }
    }

    @Override
    public void close() {
        if (closed) return;
        lock.lock();
        try {
            closed = true;
            writer.flush();
            writer.close();
        } catch (IOException e) {
            LOG.error("FileSink close failed", e);
        } finally {
            lock.unlock();
        }
    }

    /** Visible for tests. */
    public Path activeFile() {
        return activeFile;
    }
}

File: python-logger/structured_logging/sinks/file_sink.py
"""NDJSON FileSink. One envelope per line. Size-based rotation.
Designed to be tailed by a sidecar (S3BatchSink, kafka-tailer, etc.)."""
from __future__ import annotations

import logging
import os
import threading
from pathlib import Path
from typing import Optional

from .base import LogEnvelope, LogSink, SinkCallback, envelope_to_json_bytes


class FileSink(LogSink):
    def __init__(
        self,
        active_path: os.PathLike,
        *,
        rotate_bytes: int = 64 * 1024 * 1024,
        fsync_on_flush: bool = False,
    ) -> None:
        self.active = Path(active_path)
        self.active.parent.mkdir(parents=True, exist_ok=True)
        self.rotate_bytes = rotate_bytes
        self.fsync_on_flush = fsync_on_flush
        self._lock = threading.Lock()
        self._fh = open(self.active, "ab", buffering=0)
        self._bytes = self.active.stat().st_size
        self._rotation_idx = 0
        self._log = logging.getLogger("structured_logging.sink.file")

    def name(self) -> str:
        return f"file[{self.active}]"

    def publish(self, env: LogEnvelope, callback: SinkCallback = None) -> None:
        line = envelope_to_json_bytes(env) + b"\n"
        try:
            with self._lock:
                self._fh.write(line)
                self._bytes += len(line)
                if self._bytes >= self.rotate_bytes:
                    self._rotate_locked()
        except Exception as e:  # noqa: BLE001
            self._log.error("file write threw: %s", e)
            if callback:
                callback(False, e)
            return
        if callback:
            callback(True, None)

    def _rotate_locked(self) -> None:
        try:
            self._fh.close()
        except Exception:
            pass
        self._rotation_idx += 1
        rotated = Path(str(self.active) + f".{self._rotation_idx}")
        try:
            self.active.rename(rotated)
        except Exception as e:  # noqa: BLE001
            self._log.warning("rename %s -> %s failed: %s", self.active, rotated, e)
        self._fh = open(self.active, "ab", buffering=0)
        self._bytes = 0

    def flush(self, timeout_s: float = 30.0) -> None:
        with self._lock:
            try:
                self._fh.flush()
                if self.fsync_on_flush:
                    os.fsync(self._fh.fileno())
            except Exception:
                pass

    def close(self) -> None:
        with self._lock:
            try:
                self._fh.flush()
            finally:
                self._fh.close()
